lib/resty/iprepd.lua

```diff
@@ -0,0 +1,78 @@
+local cjson = require('cjson')
+local lrucache = require('resty.lrucache')
+local http = require('resty.http')
+
+function fatal_error(error_msg)
+  ngx.log(ngx.ERR, error_msg)
+  os.exit(1)
+end
+
+local _M = {}
+local mt = { __index = _M }
+
+function _M.new(options)
+  local iprepd_url = options.url or 'http://localhost:8080/'
+  local cache_ttl = options.cache_ttl or 30
+
+  local iprepd_threshold = options.threshold or fatal_error('Need to pass in a threshold')
+  local iprepd_api_key = options.api_key or fatal_error('Need to pass in an api_key')
+
+  -- TODO: Make configurable?
+  local cache, err = lrucache.new(200)
+  if not cache then
+    fatal_error('failed to create the cache: ' .. (err or 'unknown'))
+  end
+
+  local self = {
+    url = iprepd_url,
+    threshold = iprepd_threshold,
+    api_key_hdr = {
+      ['Authorization'] = 'APIKey ' .. iprepd_api_key,
+    },
+    cache_ttl = cache_ttl,
+    timeout = options.timeout or 10,
+    cache = cache,
+  }
+  return setmetatable(self, mt)
+end
+
+function _M.check(self, ip)
+  local httpc = http.new()
+  -- set timeout in ms
+  httpc:set_timeout(self.timeout)
+
+  -- Get reputation for ip
+  local reputation = self.cache:get(ip)
+  if not reputation then
+    local resp, err = httpc:request_uri(self.url .. '/' .. ip, {
+      method  = "GET",
+      headers = self.api_key_hdr,
+    })
+    if err then
+      ngx.log(ngx.ERR, 'Error with request to iprepd: ' .. err)
+    else
+      -- If the IP was found
+      if resp and resp.status == 200 then
+        reputation = cjson.decode(resp.body)['reputation']
+        if reputation and reputation >= 0 and reputation <= 100 then
+          self.cache:set(ngx.var.remote_addr, reputation, self.cache_ttl)
```

Comment:
This hard-codes the source of the ip, when it should use what was passed-in instead.

```suggestion
          self.cache:set(ip, reputation, self.cache_ttl)
```